ui/components/app/currency-input/currency-input.js

```diff
@@ -41,84 +45,93 @@ export default function CurrencyInput({
   onPreferenceToggle,
   swapIcon,
   className = '',
+  // if null, the asset is the native currency
+  asset,
 }) {
   const t = useContext(I18nContext);
 
+  const assetDecimals = asset?.decimals || NATIVE_CURRENCY_DECIMALS;
+
   const preferredCurrency = useSelector(getNativeCurrency);
   const secondaryCurrency = useSelector(getCurrentCurrency);
-  const conversionRate = useSelector(getConversionRate);
+
+  const primarySuffix =
+    asset?.symbol || preferredCurrency || EtherDenomination.ETH;
+  const secondarySuffix = secondaryCurrency.toUpperCase();
+  const isLongSymbol = (primarySuffix?.length || 0) > LARGE_SYMBOL_LENGTH;
+
+  const [shouldDisplayFiat, setShouldDisplayFiat] = useState(featureSecondary);
   const showFiat = useSelector(getShouldShowFiat);
+  const hideSecondary = !showFiat;
+  const shouldUseFiat = hideSecondary ? false : Boolean(shouldDisplayFiat);
+  const isTokenPrimary = !shouldUseFiat;
+
+  const [tokenDecimalValue, setTokenDecimalValue] = useState('0');
+  const [fiatDecimalValue, setFiatDecimalValue] = useState('0');
+
   const chainId = useSelector(getCurrentChainId);
   const { ticker, type } = useSelector(getProviderConfig);
   const isOriginalNativeSymbol = useIsOriginalNativeTokenSymbol(
     chainId,
     ticker,
     type,
   );
-  const hideSecondary = !showFiat;
-  const primarySuffix = preferredCurrency || EtherDenomination.ETH;
-  const secondarySuffix = secondaryCurrency.toUpperCase();
-
-  const [isSwapped, setSwapped] = useState(false);
-  const [newHexValue, setNewHexValue] = useState(hexValue);
-  const [shouldDisplayFiat, setShouldDisplayFiat] = useState(featureSecondary);
-  const shouldUseFiat = hideSecondary ? false : Boolean(shouldDisplayFiat);
 
-  const getDecimalValue = () => {
-    const decimalValueString = shouldUseFiat
-      ? getValueFromWeiHex({
-          value: hexValue,
-          toCurrency: secondaryCurrency,
-          conversionRate,
-          numberOfDecimals: 2,
-        })
-      : getValueFromWeiHex({
-          value: hexValue,
-          toCurrency: EtherDenomination.ETH,
-          numberOfDecimals: 8,
-        });
-
-    return Number(decimalValueString) || 0;
-  };
+  const tokenToFiatConversionRate = useTokenExchangeRate(asset?.address);
 
-  const initialDecimalValue = hexValue ? getDecimalValue() : 0;
+  const processNewDecimalValue = useProcessNewDecimalValue(
+    assetDecimals,
+    !isTokenPrimary,
```

Comment:
Nit: use the name `isTokenPrimary` in this file and in useProcessNewDecimalValue (negate value as needed) for better readability

Reply:
Good call. Done